core/net/ipv6/websocket-http-client.c

```diff
@@ -0,0 +1,337 @@
+};
+/*---------------------------------------------------------------------------*/
+static void
+send_get(struct websocket_http_client_state *s)
+{
+  struct tcp_socket *tcps;
+
+  tcps = &s->s;
+  tcp_socket_send_str(tcps, "GET ");
+  tcp_socket_send_str(tcps, s->file);
+  tcp_socket_send_str(tcps, " HTTP/1.1\r\n");
+  tcp_socket_send_str(tcps, "Host: ");
+  tcp_socket_send_str(tcps, s->host);
+  tcp_socket_send_str(tcps, "\r\n");
+  if(strlen(s->header) > 0) {
+    tcp_socket_send_str(tcps, s->header);
+  }
+  tcp_socket_send_str(tcps,
+                      "Connection: Upgrade\r\n"
+                      "Upgrade: websocket\r\n"
+                      "Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==\r\n"
+                      "Sec-WebSocket-Version: 13\r\n"
+                      "Sec-WebSocket-Protocol:");
+  tcp_socket_send_str(tcps, s->subprotocol);
+  tcp_socket_send_str(tcps, "\r\n");
+  tcp_socket_send_str(tcps, "\r\n");
+  PRINTF("websocket-http-client: send_get(): output buffer left %d\n", tcp_socket_max_sendlen(tcps));
+}
+/*---------------------------------------------------------------------------*/
+static void
+send_connect(struct websocket_http_client_state *s)
+{
+  struct tcp_socket *tcps;
+  char buf[20];
+
+  tcps = &s->s;
+  tcp_socket_send_str(tcps, "CONNECT ");
+  tcp_socket_send_str(tcps, s->host);
+  tcp_socket_send_str(tcps, ":");
+  sprintf(buf, "%d", s->port);
+  tcp_socket_send_str(tcps, buf);
+  tcp_socket_send_str(tcps, " HTTP/1.1\r\n");
+  tcp_socket_send_str(tcps, "Host: ");
+  tcp_socket_send_str(tcps, s->host);
+  tcp_socket_send_str(tcps, "\r\n");
+  tcp_socket_send_str(tcps, "Proxy-Connection: Keep-Alive\r\n\r\n");
+}
+/*---------------------------------------------------------------------------*/
+static void
+event(struct tcp_socket *tcps, void *ptr,
+      tcp_socket_event_t e)
+{
+  struct websocket_http_client_state *s = ptr;
+
+  if(e == TCP_SOCKET_CONNECTED) {
+    if(s->proxy_port != 0) {
+      send_connect(s);
+    } else {
+      send_get(s);
+    }
+  } else if(e == TCP_SOCKET_CLOSED) {
+    websocket_http_client_closed(s);
+  } else if(e == TCP_SOCKET_TIMEDOUT) {
+    websocket_http_client_timedout(s);
+  } else if(e == TCP_SOCKET_ABORTED) {
+    websocket_http_client_aborted(s);
+  } else if(e == TCP_SOCKET_DATA_SENT) {
+  }
+}
+/*---------------------------------------------------------------------------*/
+static int
+parse_header_byte(struct websocket_http_client_state *s,
+                  uint8_t b)
+{
+  static const char *endmarker = "\r\n\r\n";
+
+  PT_BEGIN(&s->parse_header_pt);
+
+  /* Skip the first part of the HTTP response */
+  while(b != ' ') {
+    PT_YIELD(&s->parse_header_pt);
+  }
+
+  /* Skip the space that follow the first part */
+  PT_YIELD(&s->parse_header_pt);
+
+  /* Read the first three bytes that consistute the HTTP status code. */
+  s->http_status = (b - '0');
+  PT_YIELD(&s->parse_header_pt);
+  s->http_status = s->http_status * 10 + (b - '0');
```

Comment:
What does this exactly does? adding a comment here would be better for readiness